test/soak/serve_hostnames/serve_hostnames.go

```diff
@@ -236,26 +245,46 @@ func main() {
 
 	// Repeatedly make requests.
 	for iteration := 0; iteration != *upTo; iteration++ {
-		responses := make(map[string]int, *podsPerNode*len(nodes.Items))
+		responseChan := make(chan string, queries)
+		// Use a channel of size *maxPar to throttle the number
+		// of in-flight requests to avoid overloading the service.
+		inFlight := make(chan struct{}, *maxPar)
 		start := time.Now()
 		for q := 0; q < queries; q++ {
-			t := time.Now()
-			hostname, err := c.Get().
-				Namespace(ns).
-				Prefix("proxy").
-				Resource("services").
-				Name("serve-hostnames").
-				DoRaw()
-			glog.V(4).Infof("Proxy call in namespace %s took %v", ns, time.Since(t))
-			if err != nil {
-				glog.Infof("Call failed during iteration %d query %d : %v", iteration, q, err)
-			} else {
-				responses[string(hostname)]++
-			}
-
+			go func() {
+				inFlight <- struct{}{}
+				t := time.Now()
+				hostname, err := c.Get().
+					Namespace(ns).
+					Prefix("proxy").
+					Resource("services").
+					Name("serve-hostnames").
+					DoRaw()
+				glog.V(4).Infof("Proxy call in namespace %s took %v", ns, time.Since(t))
+				if err != nil {
+					glog.Warningf("!Call failed during iteration %d query %d : %v", iteration, q, err)
+					responseChan <- fmt.Sprintf("failed in iteration %d to issue query %d: %v", iteration, q, err)
+				} else {
+					responseChan <- string(hostname)
+				}
+				<-inFlight
+			}()
 		}
+		responses := make(map[string]int, *podsPerNode*len(nodes.Items))
+		for q := 0; q < queries; q++ {
+			r := <-responseChan
+			glog.V(4).Infof("Got response from %s", r)
+			responses[r]++
+		}
+		missing := 0
 		for k, v := range responses {
-			glog.V(4).Infof("%s: %d  ", k, v)
+			glog.V(3).Infof("%s: %d  ", k, v)
+			if len(k) > 0 && k[0] == '!' {
```

Comment:
This won't match what you want because while you put a ! as the first char of the statement written to the log the first char of the string written to the channel when there is an error is 'f'.